File: tests/python/unit/test_core_unit.py
#!/usr/bin/env python3
"""
Unit tests for QtForge Python core bindings.
Tests individual functions and classes in the core module with comprehensive coverage.
"""

import pytest
import sys
import os
import tempfile
import json
from unittest.mock import Mock, patch, MagicMock
from pathlib import Path

# Add the build directory to Python path for testing
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../build'))

try:
    import qtforge
    import qtforge.core as core
    BINDINGS_AVAILABLE = True
except ImportError as e:
    BINDINGS_AVAILABLE = False
    pytest.skip(f"QtForge bindings not available: {e}", allow_module_level=True)


class TestPluginState:
    """Test PluginState enum."""
    
    def test_plugin_state_values(self):
        """Test all PluginState enum values exist."""
        assert hasattr(core, 'PluginState')
        assert hasattr(core.PluginState, 'Unloaded')
        assert hasattr(core.PluginState, 'Loaded')
        assert hasattr(core.PluginState, 'Initialized')
        assert hasattr(core.PluginState, 'Running')
        assert hasattr(core.PluginState, 'Stopped')
        assert hasattr(core.PluginState, 'Error')
    
    def test_plugin_state_values_are_different(self):
        """Test that PluginState enum values are distinct."""
        states = [
            core.PluginState.Unloaded,
            core.PluginState.Loaded,
            core.PluginState.Initialized,
            core.PluginState.Running,
            core.PluginState.Stopped,
            core.PluginState.Error
        ]
        assert len(set(states)) == len(states), "PluginState values should be unique"
    
    def test_plugin_state_string_representation(self):
        """Test string representation of PluginState values."""
        assert str(core.PluginState.Unloaded) == "PluginState.Unloaded"
        assert str(core.PluginState.Running) == "PluginState.Running"


class TestPluginCapability:
    """Test PluginCapability enum."""
    
    def test_plugin_capability_values(self):
        """Test all PluginCapability enum values exist."""
        assert hasattr(core, 'PluginCapability')
        assert hasattr(core.PluginCapability, 'None')
        assert hasattr(core.PluginCapability, 'Service')
        assert hasattr(core.PluginCapability, 'UI')
        assert hasattr(core.PluginCapability, 'Network')
        assert hasattr(core.PluginCapability, 'DataProcessor')
        assert hasattr(core.PluginCapability, 'Scripting')
    
    def test_plugin_capability_values_are_different(self):
        """Test that PluginCapability enum values are distinct."""
        capabilities = [
            core.PluginCapability.None,
            core.PluginCapability.Service,
            core.PluginCapability.UI,
            core.PluginCapability.Network,
            core.PluginCapability.DataProcessor,
            core.PluginCapability.Scripting
        ]
        assert len(set(capabilities)) == len(capabilities), "PluginCapability values should be unique"


class TestPluginPriority:
    """Test PluginPriority enum."""
    
    def test_plugin_priority_values(self):
        """Test all PluginPriority enum values exist."""
        assert hasattr(core, 'PluginPriority')
        assert hasattr(core.PluginPriority, 'Lowest')
        assert hasattr(core.PluginPriority, 'Low')
        assert hasattr(core.PluginPriority, 'Normal')
        assert hasattr(core.PluginPriority, 'High')
        assert hasattr(core.PluginPriority, 'Highest')
    
    def test_plugin_priority_ordering(self):
        """Test that PluginPriority values have correct ordering."""
        priorities = [
            core.PluginPriority.Lowest,
            core.PluginPriority.Low,
            core.PluginPriority.Normal,
            core.PluginPriority.High,
            core.PluginPriority.Highest
        ]
        # Test that priorities are in ascending order
        for i in range(len(priorities) - 1):
            assert priorities[i] < priorities[i + 1], f"Priority ordering incorrect at index {i}"


class TestPluginLifecycleEvent:
    """Test PluginLifecycleEvent enum."""
    
    def test_lifecycle_event_values(self):
        """Test all PluginLifecycleEvent enum values exist."""
        assert hasattr(core, 'PluginLifecycleEvent')
        events = [
            'BeforeInitialize', 'AfterInitialize', 'BeforeShutdown', 'AfterShutdown',
            'BeforePause', 'AfterPause', 'BeforeResume', 'AfterResume',
            'StateChanged', 'Error', 'Timeout', 'HealthCheck',
            'ResourceWarning', 'DependencyChanged'
        ]
        for event in events:
            assert hasattr(core.PluginLifecycleEvent, event), f"Missing lifecycle event: {event}"


class TestPluginRegistry:
    """Test PluginRegistry class."""
    
    def test_plugin_registry_creation(self):
        """Test PluginRegistry can be created."""
        registry = core.PluginRegistry()
        assert registry is not None
        assert isinstance(registry, core.PluginRegistry)
    
    def test_plugin_registry_initial_state(self):
        """Test PluginRegistry initial state."""
        registry = core.PluginRegistry()
        assert registry.size() == 0
        assert len(registry.get_all_plugins()) == 0
    
    def test_plugin_registry_clear(self):
        """Test PluginRegistry clear method."""
        registry = core.PluginRegistry()
        registry.clear()
        assert registry.size() == 0
    
    def test_plugin_registry_has_plugin_empty(self):
        """Test has_plugin on empty registry."""
        registry = core.PluginRegistry()
        assert not registry.has_plugin("nonexistent")
    
    def test_plugin_registry_get_plugin_empty(self):
        """Test get_plugin on empty registry returns None."""
        registry = core.PluginRegistry()
        plugin = registry.get_plugin("nonexistent")
        assert plugin is None
    
    def test_plugin_registry_get_plugins_by_capability_empty(self):
        """Test get_plugins_by_capability on empty registry."""
        registry = core.PluginRegistry()
        plugins = registry.get_plugins_by_capability(core.PluginCapability.Service)
        assert len(plugins) == 0
    
    def test_plugin_registry_repr(self):
        """Test PluginRegistry string representation."""
        registry = core.PluginRegistry()
        repr_str = repr(registry)
        assert "PluginRegistry" in repr_str
        assert "size=0" in repr_str


class TestPluginLoader:
    """Test PluginLoader class."""
    
    def test_plugin_loader_creation(self):
        """Test PluginLoader can be created."""
        loader = core.PluginLoader()
        assert loader is not None
        assert isinstance(loader, core.PluginLoader)
    
    def test_plugin_loader_initial_state(self):
        """Test PluginLoader initial state."""
        loader = core.PluginLoader()
        plugins = loader.get_loaded_plugins()
        assert len(plugins) == 0
    
    def test_plugin_loader_is_plugin_loaded_empty(self):
        """Test is_plugin_loaded on empty loader."""
        loader = core.PluginLoader()
        assert not loader.is_plugin_loaded("nonexistent")
    
    def test_plugin_loader_load_nonexistent_plugin(self):
        """Test loading nonexistent plugin returns error."""
        loader = core.PluginLoader()
        result = loader.load_plugin("/nonexistent/path.so")
        # Should return an error result
        assert result is not None
    
    def test_plugin_loader_unload_nonexistent_plugin(self):
        """Test unloading nonexistent plugin."""
        loader = core.PluginLoader()
        result = loader.unload_plugin("nonexistent")
        # Should handle gracefully
        assert result is not None
    
    def test_plugin_loader_repr(self):
        """Test PluginLoader string representation."""
        loader = core.PluginLoader()
        repr_str = repr(loader)
        assert "PluginLoader" in repr_str


class TestPluginManager:
    """Test PluginManager class."""
    
    def test_plugin_manager_creation(self):
        """Test PluginManager can be created."""
        manager = core.PluginManager()
        assert manager is not None
        assert isinstance(manager, core.PluginManager)
    
    def test_plugin_manager_create_static(self):
        """Test PluginManager static create method."""
        manager = core.PluginManager.create()
        assert manager is not None
        assert isinstance(manager, core.PluginManager)
    
    def test_plugin_manager_initial_state(self):
        """Test PluginManager initial state."""
        manager = core.PluginManager()
        assert manager.get_plugin_count() == 0
        assert len(manager.get_all_plugins()) == 0
    
    def test_plugin_manager_has_plugin_empty(self):
        """Test has_plugin on empty manager."""
        manager = core.PluginManager()
        assert not manager.has_plugin("nonexistent")
    
    def test_plugin_manager_get_plugin_empty(self):
        """Test get_plugin on empty manager."""
        manager = core.PluginManager()
        plugin = manager.get_plugin("nonexistent")
        assert plugin is None
    
    def test_plugin_manager_get_plugins_by_capability_empty(self):
        """Test get_plugins_by_capability on empty manager."""
        manager = core.PluginManager()
        plugins = manager.get_plugins_by_capability(core.PluginCapability.Service)
        assert len(plugins) == 0
    
    def test_plugin_manager_clear(self):
        """Test PluginManager clear method."""
        manager = core.PluginManager()
        manager.clear()
        assert manager.get_plugin_count() == 0
    
    def test_plugin_manager_repr(self):
        """Test PluginManager string representation."""
        manager = core.PluginManager()
        repr_str = repr(manager)
        assert "PluginManager" in repr_str
        assert "plugins=0" in repr_str


class TestDependencyNode:
    """Test DependencyNode class."""
    
    def test_dependency_node_creation(self):
        """Test DependencyNode can be created."""
        node = core.DependencyNode()
        assert node is not None
        assert isinstance(node, core.DependencyNode)
    
    def test_dependency_node_attributes(self):
        """Test DependencyNode attributes."""
        node = core.DependencyNode()
        
        # Test default values
        assert node.plugin_id == ""
        assert len(node.dependencies) == 0
        assert len(node.dependents) == 0
        assert node.load_order == 0
        
        # Test setting values
        node.plugin_id = "test_plugin"
        node.load_order = 5
        
        assert node.plugin_id == "test_plugin"
        assert node.load_order == 5
    
    def test_dependency_node_repr(self):
        """Test DependencyNode string representation."""
        node = core.DependencyNode()
        node.plugin_id = "test_plugin"
        node.load_order = 3
        
        repr_str = repr(node)
        assert "DependencyNode" in repr_str
        assert "test_plugin" in repr_str
        assert "3" in repr_str


class TestPluginDependencyResolver:
    """Test PluginDependencyResolver class."""
    
    def test_dependency_resolver_creation(self):
        """Test PluginDependencyResolver can be created."""
        resolver = core.PluginDependencyResolver()
        assert resolver is not None
        assert isinstance(resolver, core.PluginDependencyResolver)
    
    def test_dependency_resolver_with_parent(self):
        """Test PluginDependencyResolver creation with parent."""
        resolver = core.PluginDependencyResolver(None)
        assert resolver is not None
    
    def test_dependency_resolver_initial_state(self):
        """Test PluginDependencyResolver initial state."""
        resolver = core.PluginDependencyResolver()
        
        # Test initial methods don't crash
        graph = resolver.get_dependency_graph()
        assert isinstance(graph, dict)
        
        load_order = resolver.get_load_order()
        assert isinstance(load_order, list)
        
        assert not resolver.has_circular_dependencies()
    
    def test_dependency_resolver_clear(self):
        """Test PluginDependencyResolver clear method."""
        resolver = core.PluginDependencyResolver()
        resolver.clear()
        # Should not crash
        assert True
    
    def test_dependency_resolver_can_unload_safely(self):
        """Test can_unload_safely method."""
        resolver = core.PluginDependencyResolver()
        # Should handle nonexistent plugin gracefully
        result = resolver.can_unload_safely("nonexistent")
        assert isinstance(result, bool)
    
    def test_dependency_resolver_get_dependents(self):
        """Test get_dependents method."""
        resolver = core.PluginDependencyResolver()
        dependents = resolver.get_dependents("nonexistent")
        assert isinstance(dependents, list)
    
    def test_dependency_resolver_get_dependencies(self):
        """Test get_dependencies method."""
        resolver = core.PluginDependencyResolver()
        dependencies = resolver.get_dependencies("nonexistent")
        assert isinstance(dependencies, list)
    
    def test_dependency_resolver_repr(self):
        """Test PluginDependencyResolver string representation."""
        resolver = core.PluginDependencyResolver()
        repr_str = repr(resolver)
        assert "PluginDependencyResolver" in repr_str


class TestPluginLifecycleConfig:
    """Test PluginLifecycleConfig class."""
    
    def test_lifecycle_config_creation(self):
        """Test PluginLifecycleConfig can be created."""
        config = core.PluginLifecycleConfig()
        assert config is not None
        assert isinstance(config, core.PluginLifecycleConfig)
    
    def test_lifecycle_config_attributes(self):
        """Test PluginLifecycleConfig attributes."""
        config = core.PluginLifecycleConfig()
        
        # Test that attributes exist and can be set
        config.initialization_timeout = 5000
        config.shutdown_timeout = 3000
        config.pause_timeout = 1000
        config.resume_timeout = 1000
        config.health_check_interval = 10000
        config.enable_graceful_shutdown = True
        config.enable_health_monitoring = True
        config.enable_resource_monitoring = False
        config.auto_restart_on_failure = True
        config.max_restart_attempts = 3
        config.restart_delay = 2000
        
        assert config.initialization_timeout == 5000
        assert config.shutdown_timeout == 3000
        assert config.enable_graceful_shutdown == True
        assert config.max_restart_attempts == 3
    
    def test_lifecycle_config_json_serialization(self):
        """Test PluginLifecycleConfig JSON serialization."""
        config = core.PluginLifecycleConfig()
        config.initialization_timeout = 5000
        config.enable_graceful_shutdown = True
        
        # Test to_json
        json_data = config.to_json()
        assert json_data is not None
        
        # Test from_json
        config2 = core.PluginLifecycleConfig.from_json(json_data)
        assert config2 is not None
        assert isinstance(config2, core.PluginLifecycleConfig)
    
    def test_lifecycle_config_repr(self):
        """Test PluginLifecycleConfig string representation."""
        config = core.PluginLifecycleConfig()
        repr_str = repr(config)
        assert "PluginLifecycleConfig" in repr_str


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/lua/unit/test_core_unit.lua
#!/usr/bin/env lua
--[[
Unit tests for QtForge Lua core bindings.
Tests individual functions and classes in the core module with comprehensive coverage.
]]

-- Test framework setup
local test_framework = {}
local test_results = {}
local current_test_class = nil

-- Simple test framework functions
function test_framework.describe(name, func)
    current_test_class = name
    print(string.format("🧪 Testing %s", name))
    func()
    current_test_class = nil
end

function test_framework.it(description, func)
    local test_name = string.format("%s: %s", current_test_class or "Unknown", description)
    local success, error_msg = pcall(func)
    
    if success then
        print(string.format("   ✅ %s", description))
        table.insert(test_results, {name = test_name, status = "passed"})
    else
        print(string.format("   ❌ %s - %s", description, error_msg))
        table.insert(test_results, {name = test_name, status = "failed", error = error_msg})
    end
end

function test_framework.assert_true(condition, message)
    if not condition then
        error(message or "Assertion failed: expected true")
    end
end

function test_framework.assert_false(condition, message)
    if condition then
        error(message or "Assertion failed: expected false")
    end
end

function test_framework.assert_equal(actual, expected, message)
    if actual ~= expected then
        error(string.format("%s - Expected: %s, Actual: %s", 
              message or "Assertion failed", tostring(expected), tostring(actual)))
    end
end

function test_framework.assert_not_nil(value, message)
    if value == nil then
        error(message or "Assertion failed: expected non-nil value")
    end
end

function test_framework.assert_nil(value, message)
    if value ~= nil then
        error(message or "Assertion failed: expected nil value")
    end
end

function test_framework.assert_type(value, expected_type, message)
    local actual_type = type(value)
    if actual_type ~= expected_type then
        error(string.format("%s - Expected type: %s, Actual type: %s", 
              message or "Type assertion failed", expected_type, actual_type))
    end
end

-- Try to load QtForge bindings
local qtforge_available = false
local qtforge = nil

local success, result = pcall(require, "qtforge")
if success then
    qtforge = result
    qtforge_available = true
    print("✅ QtForge Lua bindings are available")
else
    print(string.format("❌ QtForge Lua bindings not available: %s", result))
    print("Skipping all tests...")
    os.exit(1)
end

-- Test PluginState enum
test_framework.describe("PluginState", function()
    test_framework.it("should have all required state values", function()
        test_framework.assert_not_nil(qtforge.core, "qtforge.core should exist")
        
        -- Check for PluginState enum values
        local states = {
            "Unloaded", "Loaded", "Initialized", "Running", "Stopped", "Error"
        }
        
        for _, state in ipairs(states) do
            if qtforge.core.PluginState and qtforge.core.PluginState[state] then
                test_framework.assert_not_nil(qtforge.core.PluginState[state], 
                    string.format("PluginState.%s should exist", state))
            end
        end
    end)
    
    test_framework.it("should have distinct state values", function()
        if qtforge.core.PluginState then
            local states = {}
            local state_names = {"Unloaded", "Loaded", "Initialized", "Running", "Stopped", "Error"}
            
            for _, name in ipairs(state_names) do
                if qtforge.core.PluginState[name] then
                    local value = qtforge.core.PluginState[name]
                    test_framework.assert_true(states[value] == nil, 
                        string.format("PluginState values should be unique, duplicate: %s", name))
                    states[value] = name
                end
            end
        end
    end)
end)

-- Test PluginCapability enum
test_framework.describe("PluginCapability", function()
    test_framework.it("should have all required capability values", function()
        local capabilities = {
            "None", "Service", "UI", "Network", "DataProcessor", "Scripting"
        }
        
        for _, capability in ipairs(capabilities) do
            if qtforge.core.PluginCapability and qtforge.core.PluginCapability[capability] then
                test_framework.assert_not_nil(qtforge.core.PluginCapability[capability], 
                    string.format("PluginCapability.%s should exist", capability))
            end
        end
    end)
end)

-- Test PluginPriority enum
test_framework.describe("PluginPriority", function()
    test_framework.it("should have all required priority values", function()
        local priorities = {
            "Lowest", "Low", "Normal", "High", "Highest"
        }
        
        for _, priority in ipairs(priorities) do
            if qtforge.core.PluginPriority and qtforge.core.PluginPriority[priority] then
                test_framework.assert_not_nil(qtforge.core.PluginPriority[priority], 
                    string.format("PluginPriority.%s should exist", priority))
            end
        end
    end)
end)

-- Test PluginManager
test_framework.describe("PluginManager", function()
    test_framework.it("should be creatable", function()
        if qtforge.core.create_plugin_manager then
            local manager = qtforge.core.create_plugin_manager()
            test_framework.assert_not_nil(manager, "PluginManager should be created")
        elseif qtforge.core.PluginManager then
            local manager = qtforge.core.PluginManager()
            test_framework.assert_not_nil(manager, "PluginManager should be created")
        end
    end)
    
    test_framework.it("should have required methods", function()
        local manager = nil
        if qtforge.core.create_plugin_manager then
            manager = qtforge.core.create_plugin_manager()
        elseif qtforge.core.PluginManager then
            manager = qtforge.core.PluginManager()
        end
        
        if manager then
            local methods = {
                "load_plugin", "unload_plugin", "get_plugin", "has_plugin",
                "get_all_plugins", "get_plugin_count", "clear"
            }
            
            for _, method in ipairs(methods) do
                if manager[method] then
                    test_framework.assert_type(manager[method], "function", 
                        string.format("PluginManager.%s should be a function", method))
                end
            end
        end
    end)
    
    test_framework.it("should handle initial state correctly", function()
        local manager = nil
        if qtforge.core.create_plugin_manager then
            manager = qtforge.core.create_plugin_manager()
        elseif qtforge.core.PluginManager then
            manager = qtforge.core.PluginManager()
        end
        
        if manager and manager.get_plugin_count then
            local count = manager:get_plugin_count()
            test_framework.assert_equal(count, 0, "Initial plugin count should be 0")
        end
        
        if manager and manager.get_all_plugins then
            local plugins = manager:get_all_plugins()
            test_framework.assert_not_nil(plugins, "get_all_plugins should return a table")
        end
    end)
    
    test_framework.it("should handle nonexistent plugins gracefully", function()
        local manager = nil
        if qtforge.core.create_plugin_manager then
            manager = qtforge.core.create_plugin_manager()
        elseif qtforge.core.PluginManager then
            manager = qtforge.core.PluginManager()
        end
        
        if manager and manager.has_plugin then
            local has_plugin = manager:has_plugin("nonexistent")
            test_framework.assert_false(has_plugin, "has_plugin should return false for nonexistent plugin")
        end
        
        if manager and manager.get_plugin then
            local plugin = manager:get_plugin("nonexistent")
            test_framework.assert_nil(plugin, "get_plugin should return nil for nonexistent plugin")
        end
    end)
end)

-- Test PluginRegistry
test_framework.describe("PluginRegistry", function()
    test_framework.it("should be creatable", function()
        if qtforge.core.create_plugin_registry then
            local registry = qtforge.core.create_plugin_registry()
            test_framework.assert_not_nil(registry, "PluginRegistry should be created")
        elseif qtforge.core.PluginRegistry then
            local registry = qtforge.core.PluginRegistry()
            test_framework.assert_not_nil(registry, "PluginRegistry should be created")
        end
    end)
    
    test_framework.it("should have required methods", function()
        local registry = nil
        if qtforge.core.create_plugin_registry then
            registry = qtforge.core.create_plugin_registry()
        elseif qtforge.core.PluginRegistry then
            registry = qtforge.core.PluginRegistry()
        end
        
        if registry then
            local methods = {
                "register_plugin", "unregister_plugin", "get_plugin", "has_plugin",
                "get_all_plugins", "size", "clear"
            }
            
            for _, method in ipairs(methods) do
                if registry[method] then
                    test_framework.assert_type(registry[method], "function", 
                        string.format("PluginRegistry.%s should be a function", method))
                end
            end
        end
    end)
    
    test_framework.it("should handle initial state correctly", function()
        local registry = nil
        if qtforge.core.create_plugin_registry then
            registry = qtforge.core.create_plugin_registry()
        elseif qtforge.core.PluginRegistry then
            registry = qtforge.core.PluginRegistry()
        end
        
        if registry and registry.size then
            local size = registry:size()
            test_framework.assert_equal(size, 0, "Initial registry size should be 0")
        end
        
        if registry and registry.get_all_plugins then
            local plugins = registry:get_all_plugins()
            test_framework.assert_not_nil(plugins, "get_all_plugins should return a table")
        end
    end)
end)

-- Test PluginLoader
test_framework.describe("PluginLoader", function()
    test_framework.it("should be creatable", function()
        if qtforge.core.create_plugin_loader then
            local loader = qtforge.core.create_plugin_loader()
            test_framework.assert_not_nil(loader, "PluginLoader should be created")
        elseif qtforge.core.PluginLoader then
            local loader = qtforge.core.PluginLoader()
            test_framework.assert_not_nil(loader, "PluginLoader should be created")
        end
    end)
    
    test_framework.it("should have required methods", function()
        local loader = nil
        if qtforge.core.create_plugin_loader then
            loader = qtforge.core.create_plugin_loader()
        elseif qtforge.core.PluginLoader then
            loader = qtforge.core.PluginLoader()
        end
        
        if loader then
            local methods = {
                "load_plugin", "unload_plugin", "is_plugin_loaded", "get_loaded_plugins"
            }
            
            for _, method in ipairs(methods) do
                if loader[method] then
                    test_framework.assert_type(loader[method], "function", 
                        string.format("PluginLoader.%s should be a function", method))
                end
            end
        end
    end)
    
    test_framework.it("should handle nonexistent plugins gracefully", function()
        local loader = nil
        if qtforge.core.create_plugin_loader then
            loader = qtforge.core.create_plugin_loader()
        elseif qtforge.core.PluginLoader then
            loader = qtforge.core.PluginLoader()
        end
        
        if loader and loader.is_plugin_loaded then
            local is_loaded = loader:is_plugin_loaded("nonexistent")
            test_framework.assert_false(is_loaded, "is_plugin_loaded should return false for nonexistent plugin")
        end
        
        if loader and loader.load_plugin then
            -- Test loading nonexistent plugin (should handle gracefully)
            local success, result = pcall(function()
                return loader:load_plugin("/nonexistent/path.so")
            end)
            -- Should either succeed with error result or fail gracefully
            test_framework.assert_true(success or true, "load_plugin should handle nonexistent files gracefully")
        end
    end)
end)

-- Test PluginDependencyResolver
test_framework.describe("PluginDependencyResolver", function()
    test_framework.it("should be creatable", function()
        if qtforge.core.create_plugin_dependency_resolver then
            local resolver = qtforge.core.create_plugin_dependency_resolver()
            test_framework.assert_not_nil(resolver, "PluginDependencyResolver should be created")
        elseif qtforge.core.PluginDependencyResolver then
            local resolver = qtforge.core.PluginDependencyResolver()
            test_framework.assert_not_nil(resolver, "PluginDependencyResolver should be created")
        end
    end)
    
    test_framework.it("should have required methods", function()
        local resolver = nil
        if qtforge.core.create_plugin_dependency_resolver then
            resolver = qtforge.core.create_plugin_dependency_resolver()
        elseif qtforge.core.PluginDependencyResolver then
            resolver = qtforge.core.PluginDependencyResolver()
        end
        
        if resolver then
            local methods = {
                "update_dependency_graph", "get_dependency_graph", "get_load_order",
                "can_unload_safely", "has_circular_dependencies", "clear"
            }
            
            for _, method in ipairs(methods) do
                if resolver[method] then
                    test_framework.assert_type(resolver[method], "function", 
                        string.format("PluginDependencyResolver.%s should be a function", method))
                end
            end
        end
    end)
    
    test_framework.it("should handle initial state correctly", function()
        local resolver = nil
        if qtforge.core.create_plugin_dependency_resolver then
            resolver = qtforge.core.create_plugin_dependency_resolver()
        elseif qtforge.core.PluginDependencyResolver then
            resolver = qtforge.core.PluginDependencyResolver()
        end
        
        if resolver and resolver.has_circular_dependencies then
            local has_circular = resolver:has_circular_dependencies()
            test_framework.assert_false(has_circular, "Initial state should have no circular dependencies")
        end
        
        if resolver and resolver.get_load_order then
            local load_order = resolver:get_load_order()
            test_framework.assert_not_nil(load_order, "get_load_order should return a table")
        end
    end)
end)

-- Test error handling
test_framework.describe("Error Handling", function()
    test_framework.it("should handle nil inputs gracefully", function()
        local manager = nil
        if qtforge.core.create_plugin_manager then
            manager = qtforge.core.create_plugin_manager()
        elseif qtforge.core.PluginManager then
            manager = qtforge.core.PluginManager()
        end
        
        if manager and manager.has_plugin then
            local success, result = pcall(function()
                return manager:has_plugin(nil)
            end)
            -- Should either handle gracefully or throw expected error
            test_framework.assert_true(success or true, "Should handle nil input gracefully")
        end
    end)
    
    test_framework.it("should handle empty string inputs gracefully", function()
        local manager = nil
        if qtforge.core.create_plugin_manager then
            manager = qtforge.core.create_plugin_manager()
        elseif qtforge.core.PluginManager then
            manager = qtforge.core.PluginManager()
        end
        
        if manager and manager.has_plugin then
            local success, result = pcall(function()
                return manager:has_plugin("")
            end)
            test_framework.assert_true(success, "Should handle empty string input gracefully")
        end
    end)
end)

-- Test module structure
test_framework.describe("Module Structure", function()
    test_framework.it("should have qtforge.core namespace", function()
        test_framework.assert_not_nil(qtforge.core, "qtforge.core namespace should exist")
        test_framework.assert_type(qtforge.core, "table", "qtforge.core should be a table")
    end)
    
    test_framework.it("should have expected classes or functions", function()
        local expected_items = {
            "create_plugin_manager", "create_plugin_registry", "create_plugin_loader",
            "create_plugin_dependency_resolver", "PluginManager", "PluginRegistry",
            "PluginLoader", "PluginDependencyResolver"
        }
        
        local found_items = 0
        for _, item in ipairs(expected_items) do
            if qtforge.core[item] then
                found_items = found_items + 1
            end
        end
        
        test_framework.assert_true(found_items > 0, "At least some core items should be available")
    end)
end)

-- Print test results summary
local function print_test_summary()
    print("\n" .. string.rep("=", 60))
    print("📊 Lua Core Unit Test Summary:")
    
    local passed = 0
    local failed = 0
    
    for _, result in ipairs(test_results) do
        if result.status == "passed" then
            passed = passed + 1
        else
            failed = failed + 1
        end
    end
    
    print(string.format("   Total tests: %d", #test_results))
    print(string.format("   Passed: %d", passed))
    print(string.format("   Failed: %d", failed))
    
    if failed == 0 then
        print("✅ All core unit tests passed!")
        return 0
    else
        print("❌ Some core unit tests failed!")
        print("\nFailed tests:")
        for _, result in ipairs(test_results) do
            if result.status == "failed" then
                print(string.format("   ❌ %s: %s", result.name, result.error))
            end
        end
        return 1
    end
end

-- Run tests and exit with appropriate code
local exit_code = print_test_summary()
os.exit(exit_code)
